### ISSUE TITLE
Can we add support for tvos?

### ISSUE THREAD

**Issue by variantf** [grpc/grpc#14173]
seems like we don't have set  tvos.development_target in all podspecs.

**Comment by variantf** [comment]
third-party/protobuf has already support tvos now

**Comment by Allezxandre** [comment]
I have been trying to add tvOS to gRPC-Core, but because it relies on BoringSSL, I also had to add tvOS support to BoringSSL. However, I can't get to test it. `pod spec lint` cannot work for gRPC-Core as it relies on changes on BoringSSL upstream. So instead, I'm using a spare Podfile of this shape:

```ruby
pod 'BoringSSL', :path => 'grpc/src/objective-c/BoringSSL.podspec'
pod 'gRPC-Core', :path => 'grpc/gRPC-Core.podspec'
pod 'SwiftGRPC', :path => 'grpc-swift/SwiftGRPC.podspec'
```

to make sure gRPC-Core uses the edited podspec of BoringSSL that supports tvOS, but BoringSSL fails to build at the `ss.prepare_command` trying to write the umbrella files with

```
/bin/bash: line 3: include/openssl/umbrella.h: No such file or directory
```

Anyone else working on this?

**Comment by samsidd** [comment]
@Allezxandre were you able to solve it? I am also having the same issue, when I try to run 
`pod lib lint BoringSSL.podspec` it fails here `Add a module map and an umbrella header
cat > include/openssl/umbrella.h <<EOF ... `

Do you have a version of gRPC that supports tvOS?

**Comment by bimawa** [comment]
1.  I got BoringSSL from https://github.com/google/boringssl. 
2. Inserted podspec from https://github.com/CocoaPods/Specs/blob/9543734012238c33ea018af5a4381459c13f0916/Specs/0/8/a/BoringSSL/10.0.6/BoringSSL.podspec.json.
3.  Set support tvos in podspec. 
4.  After that cocoapod compile my pods successfule.

but while compilation whole project, I got compilation error: 

```
Undefined symbols for architecture x86_64:
  "_GRPC_SHADOW_SSL_get_secure_renegotiation_support", referenced from:
      bssl::ssl_check_clienthello_tlsext(bssl::SSL_HANDSHAKE*) in t1_lib.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by muxi** [comment]
gRPC is using pod `BoringSSL-GRPC` rather than `BoringSSL`

**Comment by bimawa** [comment]
@muxi ok, thanks, I'll try connecting BoringSSL-GPRC. Currently, I have an issue on pods compilation as @Allezxandre 

```
Add a module map and an umbrella header cat > include/openssl/umbrella.h <<EOF ...
```

**Comment by bimawa** [comment]
@muxi Yes, your advice helped. Now cocoapods installed successfully with next params:

```
  pod 'BoringSSL-GRPC', :podspec => 'grpc/src/objective-c'
  pod 'gRPC-Core', :podspec => 'grpc'
  pod 'SwiftGRPC', :podspec => 'grpc-swift'
```

Now I have that error: 
```
In file included from Pods/BoringSSL-GRPC/crypto/fipsmodule/cipher/e_aes.c:67:
Pods/BoringSSL-GRPC/include/openssl/arm_arch.h:96:5: error: "unsupported ARM architecture"
#   error "unsupported ARM architecture"
    ^
```

Continue researching...

**Comment by davidben** [comment]
I am not familiar with tvOS, but I assume the devices themselves are sonr flavor of ARM and the simulator is x86_64. Given you are even getting to that file, it seems unlikely that x86_64 is involved here.

I would suggest figuring out what flavor of ARM it is and looking into why it's not being detected by arm_arch.h.

**Comment by AVVS** [comment]
@davidben its arm64 based on the settings I see

<img width="653" alt="screen shot 2018-12-07 at 9 42 53 am" src="https://user-images.githubusercontent.com/1713617/49663767-813ef480-fa04-11e8-94b9-94763714079d.png">

**Comment by AVVS** [comment]
Not sure if this is of any help, but

3rd gen apple tv is `ARMv7`
and newer ones are `ARMv8-a`

<img width="1713" alt="screen shot 2018-12-07 at 9 50 15 am" src="https://user-images.githubusercontent.com/1713617/49664045-8fd9db80-fa05-11e8-835a-4e8bfbb65590.png">
<img width="1724" alt="screen shot 2018-12-07 at 9 50 01 am" src="https://user-images.githubusercontent.com/1713617/49664046-8fd9db80-fa05-11e8-90b1-2029f7f49076.png">
<img width="1727" alt="screen shot 2018-12-07 at 9 49 55 am" src="https://user-images.githubusercontent.com/1713617/49664047-90727200-fa05-11e8-9941-84c11e533722.png">

**Comment by muxi** [comment]
Sorry for missing last communications.

@bimawa - I agree with @davidben that it looks like the architecture is not correctly detected here, but are you targeting a simulator or an actual device? I find it's weird that `arm_arch.h` is used (which usually means you are targeting an ARM device, hopefully a tvOS device) and at the same time `__ARM_ARCH__` and `__aarch64__` are not defined (see ref below). I wonder what exactly is the device being targeted.

Ref to related part of `arm_arch.h`:
```
#if !defined(__ARM_ARCH__)
  #if defined(__CC_ARM)
    #define __ARM_ARCH__ __TARGET_ARCH_ARM
    #if defined(__BIG_ENDIAN)
      #define __ARMEB__
    #else
      #define __ARMEL__
    #endif
  #elif defined(__GNUC__)
    #if defined(__aarch64__)
      #define __ARM_ARCH__ 8
      #if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
        #define __ARMEB__
      #else
        #define __ARMEL__
      #endif
      // Why doesn't gcc define __ARM_ARCH__? Instead it defines
      // bunch of below macros. See all_architectires[] table in
      // gcc/config/arm/arm.c. On a side note it defines
      // __ARMEL__/__ARMEB__ for little-/big-endian.
    #elif defined(__ARM_ARCH)
      #define __ARM_ARCH__ __ARM_ARCH
    #elif defined(__ARM_ARCH_8A__)
      #define __ARM_ARCH__ 8
    #elif defined(__ARM_ARCH_7__) || defined(__ARM_ARCH_7A__) || \
        defined(__ARM_ARCH_7R__) || defined(__ARM_ARCH_7M__) ||  \
        defined(__ARM_ARCH_7EM__)
      #define __ARM_ARCH__ 7
    #elif defined(__ARM_ARCH_6__) || defined(__ARM_ARCH_6J__) || \
        defined(__ARM_ARCH_6K__) || defined(__ARM_ARCH_6M__) ||  \
        defined(__ARM_ARCH_6Z__) || defined(__ARM_ARCH_6ZK__) || \
        defined(__ARM_ARCH_6T2__)
      #define __ARM_ARCH__ 6
    #elif defined(__ARM_ARCH_5__) || defined(__ARM_ARCH_5T__) || \
        defined(__ARM_ARCH_5E__) || defined(__ARM_ARCH_5TE__) || \
        defined(__ARM_ARCH_5TEJ__)
      #define __ARM_ARCH__ 5
    #elif defined(__ARM_ARCH_4__) || defined(__ARM_ARCH_4T__)
      #define __ARM_ARCH__ 4
    #else
      #error "unsupported ARM architecture"
    #endif
  #endif
#endif
```

**Comment by AVVS** [comment]
@muxi / @bimawa we are trying to build for the simulator first, any of these produce the same result (targeting tvos 10+):

<img width="259" alt="screen shot 2018-12-07 at 10 19 02 am" src="https://user-images.githubusercontent.com/1713617/49665381-8ce0ea00-fa09-11e8-9477-2046bce80635.png">

**Comment by davidben** [comment]
If it's the simulator, I assume that means you are trying to compile for x86_64. In that case you should look into why arm_arch.h is getting pulled in, because it should only be pulled in on ARM.

We build BoringSSL for the iOS simulator fine in Chromium, so I'm guessing there is something funny about how you are building it.

**Comment by bimawa** [comment]
@davidben @muxi @AVVS 
Guys thanks for your advice, I tried to reproduce the same error but after the update all of dependence source codes and manipulation with code in sources I compiled it successfully for x86-64 and for arm64 too. 
So my worked podfile below:
```
#Test grpc for tv-os
pod 'gRPC-Core', :git => 'git@github.com:bimawa/grpc.git', :branch => 'feature/supportTVOS'
pod 'BoringSSL-GRPC', :podspec => 'https://raw.githubusercontent.com/bimawa/boringssl/feature/supportTVOS/BoringSSL-GRPC.podspec'
pod 'SwiftGRPC', :git => 'git@github.com:bimawa/grpc-swift.git', :branch => 'feature/supportTVOS'
pod 'SwiftProtobuf', :git => 'git@github.com:bimawa/swift-protobuf.git', :branch => 'feature/supportTVOS'
```
And also updated all repos in my forks. Pls. check it, and correct. Maybe can we do something with it for support tvos in swift-gRPC

**Comment by muxi** [comment]
@bimawa Thanks! Will do.

**Comment by jlaws** [comment]
@bimawa have you submitted individual PR's to fix each of those dependencies you listed? 

Getting grpc working for tvos is also blocking [Firebase on tvos](https://github.com/firebase/firebase-ios-sdk/issues/10)

**Comment by bimawa** [comment]
@jlaws No, I haven’t. I found correct commits in my forks for each dependency and did some micro fixes and GRPC start to work for tvos. But Maintainers have to take my research result and I hope they be do work for support tvos.
because I haven’t some access to BoringSSL, and other nuance did’t give to me do PR..
Right @muxi ;) ?

**Comment by bimawa** [comment]
@muxi I do some PR's. But... 
Where can I save 'https://raw.githubusercontent.com/bimawa/boringssl/feature/supportTVOS/BoringSSL-GRPC.podspec'
file for our grpc library?

And I don't sure about Apple lib  apple/swift-protobuf#833 so I wanna help to do it correctly.

**Comment by muxi** [comment]
Thanks for the PRs.

>Where can I save 'https://raw.githubusercontent.com/bimawa/boringssl/feature/supportTVOS/BoringSSL-GRPC.podspec'
file for our grpc library?

The podspec that gets uploaded is in [src/objective-c](https://github.com/grpc/grpc/tree/master/src/objective-c), but in gRPC it is generated from a template: [template/src/objective-c/BoringSSL-GRPC.podspec.template](https://github.com/grpc/grpc/blob/master/templates/src/objective-c/BoringSSL-GRPC.podspec.template).

**Comment by bimawa** [comment]
@muxi in my BoringSSL podspec 'prepare_commmand' has different body script. I think it's very important. Because I spent a lot of hours to find this worked version. How can we check it? Just I haven't skilled enough to check this body.
And second question, how can I compile templates for check it?

**Comment by muxi** [comment]
`prepare_command` section is also in the template ([here](https://github.com/grpc/grpc/blob/master/templates/src/objective-c/BoringSSL-GRPC.podspec.template#L153)). You can overwrite it.

Files are generated from templates with `$ tools/buildgen/generate_projects.sh`. I just ran it once for you in the PR you made.

**Comment by bimawa** [comment]
Ok, guys, @muxi merged it into master. Until Release you have to use 

```
  pod 'BoringSSL-GRPC', :podspec => 'https://raw.githubusercontent.com/bimawa/boringssl/feature/supportTVOS/BoringSSL-GRPC.podspec'
```

and for swift-grpc:

```
  pod 'BoringSSL-GRPC', :podspec => 'https://raw.githubusercontent.com/bimawa/boringssl/feature/supportTVOS/BoringSSL-GRPC.podspec'
```

After release, we will merge this podspecs too.

**Comment by enricmacias** [comment]
This thread saved my day! 
There are any chances on having this released anytime soon?

Thanks!

**Comment by srini100** [comment]
This should be available in 1.19.0 released a few days back.

**Comment by bimawa** [comment]
@muxi if you have some spare time, could you please help with vendoring libs in https://github.com/grpc/grpc-swift/pull/358
